File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.py
import heapq

def solution(n, paths, gates, summits):
    # 그래프 구성
    graph = [[] for _ in range(n + 1)]
    for i, j, w in paths:
        graph[i].append((j, w))
        graph[j].append((i, w))
    
    # 세트로 변환하여 O(1) 시간 복잡도로 검색 가능
    summit_set = set(summits)
    gate_set = set(gates)
    
    # 모든 출입구에서 동시에 시작하는 다익스트라 알고리즘
    INF = int(1e9)
    intensity = [INF] * (n + 1)
    q = []
    
    # 모든 출입구를 시작점으로 큐에 넣기
    for gate in gates:
        intensity[gate] = 0
        heapq.heappush(q, (0, gate))
    
    # 가장 작은 산봉우리와 강도 추적
    min_intensity = INF
    min_summit = 0
    
    # 다익스트라 알고리즘 실행
    while q:
        intens, now = heapq.heappop(q)
        
        # 현재 경로가 이미 찾은 최소 강도보다 크면 더 이상 탐색할 필요 없음
        if intens > min_intensity:
            continue
            
        # 현재 지점의 강도가 이미 계산된 강도보다 크면 스킵
        if intens > intensity[now]:
            continue
        
        # 산봉우리에 도달한 경우
        if now in summit_set:
            if intens < min_intensity or (intens == min_intensity and now < min_summit):
                min_intensity = intens
                min_summit = now
            continue  # 산봉우리에서 더 이상 탐색하지 않음
        
        # 다음 지점 탐색
        for next_node, next_intens in graph[now]:
            # 출입구로는 가지 않음
            if next_node in gate_set:
                continue
                
            # 경로의 강도는 현재까지의 최대 강도
            new_intens = max(intens, next_intens)
            
            # 더 좋은 경로를 찾은 경우에만 업데이트
            if new_intens < intensity[next_node]:
                intensity[next_node] = new_intens
                heapq.heappush(q, (new_intens, next_node))
    
    return [min_summit, min_intensity]

File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.cpp
#include <string>
#include <vector> 
#include <queue>
#include <memory.h>
#include <iostream>

using namespace std;

const int INF = 10000001;
int len;
vector<vector<pair<int, int>>> graph;
bool isSummit[50001], isGate[50001];

struct cmp {
    bool operator()(pair<int,int> a, pair<int,int> b) {
        if (a.second == b.second) return a.first > b.first;
        return a.second > b.second;
    }
};

vector<int> dijkstra(int gate) {
    // pair: node, gate부터 node까지의 최소 intensity
    priority_queue<pair<int,int>, vector<pair<int,int>>, cmp> pq;
    pq.push(make_pair(gate, 0));
    
    vector<int> intensity(len + 1, INF);
    intensity[gate] = 0;
    
    while (!pq.empty()) {
        pair<int, int> t = pq.top();
        int cur = t.first, curIntensity = t.second;
        pq.pop();
        if (isSummit[cur]) break;
        
        for (auto &edge : graph[cur]) {
            int next = edge.first, nextIntensity = edge.second;
            if (isGate[next]) continue;
            if (intensity[next] > max(curIntensity, nextIntensity)) {
                intensity[next] = max(curIntensity, nextIntensity);
                pq.push(make_pair(next, intensity[next]));
            }
        }
    }
    
    return intensity;
}

vector<int> solution(int n, vector<vector<int>> paths, vector<int> gates, vector<int> summits) {
    len = n;
    for (auto &summit : summits) {
        isSummit[summit] = true;
    }
    for (auto &gate : gates) {
        isGate[gate] = true;
    }
    
    // 1. 그래프 만들기
    graph = vector<vector<pair<int, int>>>(len + 1);
    for (auto &path : paths) {
        int from = path[0];
        int to = path[1];
        int cost = path[2];
        
        graph[from].emplace_back(to, cost);
        graph[to].emplace_back(from, cost);
    }
    
    // 2. 출입구 순회
    int minIntensity = INF;
    int prevSummit = 50001;
    vector<int> answer(2);
    for (auto &gate : gates) {
        vector<int> intensity = dijkstra(gate);
        for (auto &summit : summits) {
            if (minIntensity == intensity[summit] && prevSummit > summit) {
                answer[0] = summit;
                prevSummit = summit;
            }
            if (minIntensity > intensity[summit]) {
                minIntensity = min(minIntensity, intensity[summit]);
                answer[0] = summit;
                answer[1] = minIntensity;
                prevSummit = summit;
            }
        }
    }

    return answer;
}